File: react-native/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/PlacementConfigData.kt
package com.storylyplacementreactnative.common.data

import com.appsamurai.storyly.core.config.placement.STRLayoutDirection
import com.appsamurai.storyly.core.config.placement.STRPlacementConfig
import com.appsamurai.storyly.core.config.placement.STRProductConfig


fun decodeSTRPlacementConfig(config: Map<String, Any?>, token: String): STRPlacementConfig {
    val builder = STRPlacementConfig.Builder()

    (config["testMode"] as? Boolean)?.let { builder.setTestMode(it) }
    (config["locale"] as? String)?.let { builder.setLocale(it) }
    (config["layoutDirection"] as? String)?.let { layoutDir ->
        builder.setLayoutDirection(
            if (layoutDir == "rtl") STRLayoutDirection.RTL else STRLayoutDirection.LTR
        )
    }
    (config["customParameter"] as? String)?.let { builder.setCustomParameter(it) }
    (config["labels"] as? List<*>)?.filterIsInstance<String>()?.let { labels ->
        builder.setLabels(labels.toSet())
    }
    (config["userProperties"] as? Map<String, String>)?.let { userProps ->
        builder.setUserProperties(userProps)
    }
    (config["productConfig"] as? Map<String, Any?>)?.let { productConfig ->
        builder.setProductConfig(decodeSTRProductConfig(productConfig))
    }
    return builder.build(token)
}

fun decodeSTRProductConfig(productConfig: Map<*, *>?): STRProductConfig {
    val builder = STRProductConfig.Builder()

    if (productConfig != null) {
        (productConfig["isCartEnabled"] as? Boolean)?.let { builder.setCartAvailability(it) }
        (productConfig["isFallbackEnabled"] as? Boolean)?.let { builder.setFallbackAvailability(it) }
    }

    return builder.build()
}

File: react-native/storyly-placement-react-native/ios/Common/Data/PlacementConfigData.swift
import Foundation
import StorylyPlacement
import StorylyCore

func decodeSTRPlacementConfig(_ config: [String: Any], token: String) -> STRPlacementConfig {
    let builder = STRPlacementConfig.Builder()
    
    if let testMode = config["testMode"] as? Bool {
      builder.setTestMode(isTest: testMode)
    }
    
    if let locale = config["locale"] as? String {
      builder.setLocale(locale: locale)
    }
    
    if let layoutDirection = config["layoutDirection"] as? String {
      builder.setLayoutDirection(direction: layoutDirection == "rtl" ? .RTL : .LTR)
    }
    
    if let customParameter = config["customParameter"] as? String {
      builder.setCustomParameter(parameter: customParameter)
    }
    
    if let labels = config["labels"] as? [String] {
      builder.setLabels(labels: Set(labels))
    }
    
    if let userProperties = config["userProperties"] as? [String: String] {
      builder.setUserProperties(data: userProperties)
    }
    
    if let productConfigDict = config["productConfig"] as? [String: Any],
       let productConfig = decodeSTRProductConfig(productConfigDict) {
      builder.setProductConfig(config: productConfig)
    }
    
    if let shareConfigDict = config["shareConfig"] as? [String: Any],
       let shareConfig = decodeSTRShareConfig(shareConfigDict) {
      builder.setShareConfig(config: shareConfig)
    }
    
    if let networkConfigDict = config["networkConfig"] as? [String: Any],
       let networkConfig = decodeSTRNetworkConfig(networkConfigDict) {
      builder.setNetworkConfig(config: networkConfig)
    }
    
  return builder.build(token: token)
}

func decodeSTRProductConfig(_ productConfig: [String: Any]) -> STRProductConfig? {
    let builder = STRProductConfig.Builder()
    
    if let isCartEnabled = productConfig["isCartEnabled"] as? Bool {
        builder.setCartEnabled(isEnabled: isCartEnabled)
    }
    
    if let isFallbackEnabled = productConfig["isFallbackEnabled"] as? Bool {
        builder.setFallbackAvailability(isEnabled: isFallbackEnabled)
    }
    
    return builder.build()
}

func decodeSTRShareConfig(_ shareConfig: [String: Any]) -> STRShareConfig? {
    let builder = STRShareConfig.Builder()
    
    if let facebookAppId = shareConfig["facebookAppId"] as? String {
        builder.setFacebookAppID(id: facebookAppId)
    }
    
    if let shareUrl = shareConfig["shareUrl"] as? String {
        builder.setShareUrl(url: shareUrl)
    }
    
    if let appLogoVisibility = shareConfig["appLogoVisibility"] as? Bool {
        builder.setAppLogoVisibility(isVisible: appLogoVisibility)
    }
    
    return builder.build()
}

func decodeSTRNetworkConfig(_ networkConfig: [String: Any]) -> STRNetworkConfig? {
    let builder = STRNetworkConfig.Builder()
    
    if let cdnHost = networkConfig["cdnHost"] as? String {
        builder.setCdnHost(url: cdnHost)
    }
    
    if let productHost = networkConfig["productHost"] as? String {
        builder.setProductHost(url: productHost)
    }
    
    if let analyticHost = networkConfig["analyticHost"] as? String {
        builder.setAnalyticHost(url: analyticHost)
    }
    
    if let placementHost = networkConfig["placementHost"] as? String {
        builder.setListHost(url: placementHost)
    }
    
    return builder.build()
}

